src/server/cluster/cluster_family.cc

```diff
@@ -178,25 +178,53 @@ void ClusterFamily::ClusterInfo(ConnectionContext* cntx) {
   auto append = [&msg](absl::AlphaNum a1, absl::AlphaNum a2) {
     absl::StrAppend(&msg, a1, ":", a2, "\r\n");
   };
-  // info command just return some stats about this instance
-  int known_nodes = 1;
+
+  // Initialize response variables to emulated mode.
+  string_view state = "ok"sv;
+  SlotId slots_assigned = ClusterConfig::kMaxSlotNum + 1;
+  size_t known_nodes = 1;
   long epoch = 1;
-  ServerState& etl = *ServerState::tlocal();
-  if (etl.is_master) {
-    auto vec = server_family_->GetDflyCmd()->GetReplicasRoleInfo();
-    if (!vec.empty()) {
-      known_nodes = 2;
+  size_t cluster_size = 1;
+
+  if (is_emulated_cluster_) {
+    ServerState& etl = *ServerState::tlocal();
+    if (etl.is_master) {
+      auto vec = server_family_->GetDflyCmd()->GetReplicasRoleInfo();
+      if (!vec.empty()) {
+        known_nodes = 2;
+      }
+    } else {
+      if (server_family_->HasReplica()) {
+        known_nodes = 2;
+        epoch = server_family_->GetReplicaInfo().master_last_io_sec;
+      }
     }
+  } else if (!cluster_config_->IsConfigured()) {
+    state = "fail"sv;
+    slots_assigned = 0;
+    cluster_size = 0;
+    known_nodes = 0;
   } else {
-    if (server_family_->HasReplica()) {
-      known_nodes = 2;
-      epoch = server_family_->GetReplicaInfo().master_last_io_sec;
+    absl::flat_hash_set<string_view> nodes;
```

Comment:
you are using the sets just for counting the nodes and the cluster size.
Why not just create a counter and increase it in the for loop?

Reply:
I use the set because I did not want to assume that entries in `config` do not repeat, i.e. that a single master (for example) can appear in 2 elements of the vector.
Otherwise we can indeed use counters as you suggested.

Reply:
I am not sure we need to support this, maybe we should enforce each node to be sent only once.
If we do allow this kind of config this can lead to sending master1 with replica1 and replica2, and then master1 with replica1 only, more places for errors I think

Reply:
Indeed it complicates things, probably not worth it, at least not at the moment.
How about this:
1. I'll use simple counters for this code
2. In a separate PR, I'll add checks so that each instance appear in exactly one entry in `SetConfig()`
SG?

Reply:
sure